0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,237 @@
+    3. The remainder of the PSK data after this point will be encrypted using AES-256-CTR with the payment encryption key. Note: The ciphertext is raw binary data, and is not base64 encoded.
+    4. The sender appends the [private headers](#private-headers), followed by `\n\n`.
+    5. The sender appends the application data (in its raw binary format).
+4. The sender creates the condition of their payment by taking the HMAC of the full ILP packet with the pre-shared secret key, and computing the SHA-256 hash of it.
+5. The sender quotes and sends their payment, setting the data field of the ILP packet to the PSK data.
+6. The receiver gets notified by their ledger of an incoming prepared transfer with an ILP packet.
+7. The receiver generates the payment encryption key from the HMAC of the nonce contained in the PSK datas' public headers with the pre-shared secret key, and uses it to decrypt the private headers and their data.
+8. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call a backend to make sure the incoming funds are expected.
+9. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the same HMAC key as the sender, derived from the shared secret.
+
+## Data Format
+
+The PSK details are divided into two sets of headers, like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Application data is appended to the private headers, after a blank line. This object is then encrypted and appended to the public headers as binary data, after a blank line. Both private and public headers are parsed in the exact same way as HTTP headers.
+
+### Public Headers
+
+The public headers and their data are formatted as below:
+
+```http
+PSK/1.0
+Key: hmac-sha-256 fpwpAhlN588
+Header: data_everyone_can_see
+
+...
+```
+
+The public headers come after a single status line, which must be `PSK/1.0`. Public headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the private headers.
+
+PSK payments use an HMAC of the PSK details as a fulfillment, in order to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The public headers SHOULD only contain details required for the receiver to decrypt the private headers. The decryption key is generated as the HMAC-SHA-256 digest of the pre-shared secret key and the nonce contained in the public Key header (`fpwpAhlN588` in the above example).
```

Comment:
I would drop the first sentence here ("The public headers SHOULD...") since it's redundant with the first paragraph of this section.